src/main/java/com/metamx/common/parsers/JSONPathParser.java

```diff
@@ -0,0 +1,177 @@
+/*
+ * Licensed to Metamarkets Group Inc. (Metamarkets) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  Metamarkets licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.metamx.common.parsers;
+
+import com.jayway.jsonpath.Configuration;
+import com.jayway.jsonpath.JsonPath;
+import com.jayway.jsonpath.Option;
+import com.jayway.jsonpath.spi.json.JacksonJsonProvider;
+import com.jayway.jsonpath.spi.json.JsonProvider;
+import com.jayway.jsonpath.spi.mapper.JacksonMappingProvider;
+import com.jayway.jsonpath.spi.mapper.MappingProvider;
+import com.metamx.common.Pair;
+
+import java.util.EnumSet;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+public class JSONPathParser implements Parser<String, Object>
+{
+  private Map<String, Pair<FieldType, JsonPath>> fieldPathMap;
+  private List<FieldSpec> fieldSpecs;
+  private final boolean useFieldDiscovery;
+
+  public JSONPathParser(List<FieldSpec> fieldSpecs, boolean useFieldDiscovery)
+  {
+    this.fieldSpecs = fieldSpecs;
+    this.fieldPathMap = generateFieldPaths(fieldSpecs);
+    this.useFieldDiscovery = useFieldDiscovery;
+    setJacksonProvider();
+  }
+
+  @Override
+  public List<String> getFieldNames()
+  {
+    return null;
+  }
+
+  @Override
+  public void setFieldNames(Iterable<String> fieldNames)
+  {
+  }
+
+  @Override
+  public Map<String, Object> parse(String input)
+  {
+    Map<String, Object> map = new LinkedHashMap<>();
+    Map<String, Object> document = (Map<String, Object>) Configuration.defaultConfiguration().jsonProvider().parse(input);
+    for (Map.Entry<String, Pair<FieldType, JsonPath>> entry : fieldPathMap.entrySet()) {
+      String fieldName = entry.getKey();
+      Pair<FieldType, JsonPath> pair = entry.getValue();
+      JsonPath path = pair.rhs;
+      Object parsedVal;
+      if (pair.lhs == FieldType.ROOT) {
+        parsedVal = document.get(fieldName);
+      } else {
+        parsedVal = path.read(document);
```

Comment:
Hmm, if this works on normal `Map`s then there is actually no reason to do any of the JsonProvider business. You could just use an `ObjectMapper` directly to create the `Map` and skip the indirection.